### PR TITLE
Feature: Add Types and validations to env variables

### PR DESCRIPTION
**Pull request by timDeHof**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.
This PR adds types and validations to the env variables by using the nestjs's config package and joi package for validations. This will allow for us to group related configuration settings by function (e.g., database-related settings), and to store related settings in individual files to help manage them independently.

Two things I am unsure about are the following: 
- [x] using the config module along with 0auth providers and strategies
- [x] using the config module with the prisma schema since that's where it would need to go.

## Interesting links

[nestjs-prisma docs](https://nestjs-prisma.dev/docs/installation/)
[nestjs custom providers](https://docs.nestjs.com/fundamentals/custom-providers)
[NestJS Authentication with OAuth2.0: Configuration and Operations](https://dev.to/tugascript/nestjs-authentication-with-oauth20-configuration-and-operations-41k)
## Issue link

Fixes [#86b1k8kn6](https://app.clickup.com/t/86b1k8kn6)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Feature updates / changes
- [ ] Tests
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Run the following commands:
1. yarn test or yarn test:docker
2. yarn start:dev and test some of the endpoints in swagger
   a.Test login with discord

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have updated the change log

### CODE REVIEW

**Review by cherylli — COMMENTED** [review]
Just had a quick look and I'm not sure about some of the file placement. 

Is there any articles I should read before reviewing this PR?

File: prisma/seed/users.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { hashPassword } from "../../src/utils/auth";
+import { hashPassword } from "../../shared/auth/auth";
```
[the remark this answers is not in the document]
Comment (timDeHof) [reply prisma/seed/users.ts:1]: The reason for having a shared folder outside of `src` folder was to have a simpler method of the utils functions both inside and outside of NestJS contexts. But it does lose some of the benefits of NestJS's dependency injection. It's mainly for the seed files in the outside prisma folder.

File: src/config/mail/mail.config.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { registerAs } from "@nestjs/config";
+
+export default registerAs("mail", () => ({
+    mailjetApiPublic: process.env.MJ_APIKEY_PUBLIC,
+    mailjetApiPrivate: process.env.MJ_APIKEY_PRIVATE,
+    frontendUrl: process.env.FRONTEND_URL,
```
Comment (cherylli) [inline src/config/mail/mail.config.ts:6]: should frontendUrl be in the app config?

File: src/utils/emails/sendEmail.ts
Comment (cherylli) [inline src/utils/emails/sendEmail.ts:1]: should we rename this to something like email.service.ts?

**Review by cherylli — COMMENTED** [review]
Thanks. This is really nice.

File: prisma/production-seed/users.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import { prisma } from "../seed/prisma-client";
-import { hashPassword } from "../../src/utils/auth";
+import { hashPassword } from "../../src/global/auth/auth";
```
[the remark this answers is not in the document]
Comment (timDeHof) [reply prisma/production-seed/users.ts:2]: you are right, it might be better to name it something that allows other developers to tell what it is. I do like `auth/utils`, what do you think?

File: src/auth/auth.module.ts [hunk 4]
```diff
@@ -7,19 +7,34 @@ import { LocalStrategy } from "./strategies/local.strategy";
...
         UsersModule,
         PassportModule,
-        JwtModule.register({
-            secret: process.env.JWT_SECRET,
+        MailConfigModule,
+        JwtModule.registerAsync({
+            imports: [AuthConfigModule],
+            useFactory: async (authConfig: AuthConfig) => ({
+                secret: authConfig.secrets.JWT_SECRET,
+                signOptions: { expiresIn: "1d" },
```
[the remark this answers is not in the document]
Comment (timDeHof) [reply src/auth/auth.module.ts:31]: I think that might be something copilot suggested. I'll remove it.

File: src/config/mail/mail.config.ts (hunk 2 above)
Comment (timDeHof) [reply src/config/mail/mail.config.ts:6]: I didn't put it in the app config because it was only being used for emails and I was trying to group the config into their usage.

File: src/utils/emails/sendEmail.ts
Comment (timDeHof) [reply src/utils/emails/sendEmail.ts:1]: yes, that would probably be a better name and follow the nestjs naming convention that we have.

File: prisma/production-seed/users.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (cherylli) [reply prisma/production-seed/users.ts:2]: yeah sounds good

File: src/config/mail/mail.config.ts (hunk 2 above)
Comment (cherylli) [reply src/config/mail/mail.config.ts:6]: I think I also use that for OAuth, redirect to frontend
image

Comment (timDeHof) [reply src/config/mail/mail.config.ts:6]: ok, I'll work on moving it to appconfig then.

File: src/auth/auth.module.ts [hunk 5]
```diff
@@ -9,7 +9,7 @@ import { AtStrategy } from "./strategies/at.strategy";
 import { RtStrategy } from "./strategies/rt.strategy";
 import { DiscordStrategy } from "./strategies/discord.strategy";
 import { DiscordAuthService } from "./discord-auth.service";
-import { EmailService } from "src/utils/emails/sendEmail";
+import { EmailService } from "@/utils/emails/email.service";
```
Comment (cherylli) [inline src/auth/auth.module.ts:12]: these `@` are causing issue in e2e tests, they can only read relative paths🥲 
We can possibly fix it, `@/` looks a lot nicer, but it can be for another PR
https://stackoverflow.com/questions/64028100/nestjs-jest-cannot-find-module-with-absolute-path

**Review by cherylli — COMMENTED** [review]
e2e tests failing because of import paths

File: src/auth/auth.controller.spec.ts [hunk 6]
```diff
@@ -1,10 +1,12 @@
 import { Test, TestingModule } from "@nestjs/testing";
 import { AuthController } from "./auth.controller";
 import { AuthService } from "./auth.service";
-
+import { AppConfigService } from "@/config/app/appConfig.service";
```
Comment (cherylli) [inline src/auth/auth.controller.spec.ts:4]: this one needs to be fixed too - getting error when running e2e tests

**Comment by JoshuaHinman** [comment]
> e2e tests failing because of import paths

I ran into the same issue on line 4 here:
image

**Review by cherylli — APPROVED** [review]
working locally and all tests passed

**Review by JoshuaHinman — APPROVED** [review]
Ok - all tests passed, discord works, tried various routes in swagger - looks good!

File: prisma/seed/users.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (timDeHof) [reply prisma/seed/users.ts]: I have moved the content in the shared folder into `src/global`.

File: src/config/mail/mail.config.ts (hunk 2 above)
Comment (timDeHof) [reply src/config/mail/mail.config.ts]: I have moved `FrontendUrl` to the appConfig.

File: src/utils/emails/sendEmail.ts
Comment (timDeHof) [reply src/utils/emails/sendEmail.ts:1]: I have renamed it to follow the naming conventions we use.

**Review by cherylli — APPROVED** [review]
re-run tests after the latest changes - all passed

**Review by JoshuaHinman — APPROVED** [review]
Tests passed - looks good
